c/src/space.rs

```diff
@@ -1,12 +1,376 @@
+            if let SpaceEvent::Add(atom) = event {
+                atom_into_ptr(atom.clone())
+            } else {
+                panic!("SpaceEvent wasn't an Add event")
+            }
+        },
+        space_event_field_t::SPACE_EVENT_FIELD_REMOVE => {
+            if let SpaceEvent::Remove(atom) = event {
+                atom_into_ptr(atom.clone())
+            } else {
+                panic!("SpaceEvent wasn't a Remove event")
+            }
+        },
+        space_event_field_t::SPACE_EVENT_FIELD_REPLACE_PATTERN |
+        space_event_field_t::SPACE_EVENT_FIELD_REPLACE_TEMPLATE => {
+            if let SpaceEvent::Replace(pattern_atom, template_atom) = event {
+                match field {
+                    space_event_field_t::SPACE_EVENT_FIELD_REPLACE_PATTERN => atom_into_ptr(pattern_atom.clone()),
+                    space_event_field_t::SPACE_EVENT_FIELD_REPLACE_TEMPLATE => atom_into_ptr(template_atom.clone()),
+                    _ => unreachable!()
+                }
+            } else {
+                panic!("SpaceEvent wasn't a Replace event")
+            }
+        }
+    }
+}
+
+/// Returned space_event_t must be freed with space_event_free
+/// 
+/// WARNING: This function takes ownership of the supplied atom, so it should not be freed or
+/// accessed subsequently
+#[no_mangle]
+pub extern "C" fn space_event_new_add(atom: *mut atom_t) -> *mut space_event_t {   
+    let event = SpaceEvent::Add(ptr_into_atom(atom));
+    Box::into_raw(Box::new(space_event_t{ event }))    
+}
+
+/// Returned space_event_t must be freed with space_event_free
+/// 
+/// WARNING: This function takes ownership of the supplied atom, so it should not be freed or
+/// accessed subsequently
+#[no_mangle]
+pub extern "C" fn space_event_new_remove(atom: *mut atom_t) -> *mut space_event_t {   
+    let event = SpaceEvent::Remove(ptr_into_atom(atom));
+    Box::into_raw(Box::new(space_event_t{ event }))    
+}
+
+/// Returned space_event_t must be freed with space_event_free
+/// 
+/// WARNING: This function takes ownership of both supplied atoms, so they should not be freed or
+/// accessed subsequently
+#[no_mangle]
+pub extern "C" fn space_event_new_replace(pattern: *mut atom_t, tmpl: *mut atom_t) -> *mut space_event_t {   
+    let event = SpaceEvent::Replace(ptr_into_atom(pattern), ptr_into_atom(tmpl));
+    Box::into_raw(Box::new(space_event_t{ event }))    
+}
+
+#[no_mangle]
+pub extern "C" fn space_event_free(event: *mut space_event_t) {
+    let event = unsafe{ Box::from_raw(event) };
+    drop(event);
+}
+
+/// A handle to a space observer.
+/// 
+//QUESTION FOR VITALY: Is the idea that the same observer will observe many spaces, including
+// both C spaces & native Rust spaces?  Personally, it seems like it might be a better
+// structure to put the internal mutability, ie. Rc<RefCell<T>> pattern, inside the observer,
+// object itself, rather than on the register interface.  But perhaps I am missing something?
```

Comment:
Agh!  After spending 30m implementing and testing the change to move the RefCell inside the Observer implementations, I realized the way the code is currently structures is actually a good match for the desired semantics.  i.e. simple implementation of a SpaceObserver, and "automatic unregistration".

My objection was that I feel like exposing a RefCell in a public API is nasty, and I'm not crazy about exposing Rc either, but Rc doesn't bother me as much because many other people do it.

However, if we provide a thin wrapper with borrow() semantics, that I think it would be quite nice.  Also most of the implementation can move into the shared code for https://github.com/trueagi-io/hyperon-experimental/issues/287

I am thinking of an interface that looks like this:
```rust
/// A reference to a SpaceObserver that has been registered with a Space
#[derive(Clone)]
pub struct SpaceObserverRef<T: SpaceObserver> (Rc<T>);

impl<T> SpaceObserverRef<T> {
    pub fn borrow(&self) -> Ref<T> {
        self.0.borrow()
    }
    ...borrow_mut()
}

pub trait Space {
    ...
    fn register_observer(&self, observer: Box<dyn SpaceObserver>) -> SpaceObserverRef {
    }
```

I think this change can wait until https://github.com/trueagi-io/hyperon-experimental/issues/287 Because it's basically part of the same reorganization.

Reply:
Yes, lets discuss it further in #287